client/context/helpers.go

```diff
@@ -44,22 +51,79 @@ func (ctx CoreContext) BroadcastTx(tx []byte) (*ctypes.ResultBroadcastTxCommit,
 
 // Query from Tendermint with the provided key and storename
 func (ctx CoreContext) Query(key cmn.HexBytes, storeName string) (res []byte, err error) {
-	return ctx.query(key, storeName, "key")
+	resp, err := ctx.query(key, storeName, "key")
+	if err != nil {
+		return
+	}
+	if ctx.TrustNode {
+		return
+	}
+
+	cdc := wire.NewCodec()
+	merkle.RegisterWire(cdc)
+	store.RegisterWire(cdc)
+
+	proof, err := merkle.DecodeProof(cdc, resp.Proof, store.WrapOpDecoder(merkle.DefaultOpDecoder))
+	if err != nil {
+		return
+	}
+
+	source := liteclient.NewHTTPProvider(ctx.NodeURI)
+	trusted := lite.NewCacheProvider(lite.NewMemStoreProvider(), files.NewProvider(ctx.ProviderPath))
+
+	genesis, err := tmtypes.GenesisDocFromFile(ctx.GenesisFile)
```

Comment:
doesn't seem right that we should start over from genesis... lets hold on this, I'll figure out what we need to do to here...

Reply:
Inquiringcertifier stores retrieved headers into the trusted provider (file system). After the first time it is created it starts from the latest known header. At the first time we should start from the genesis.

Reply:
yeah but the Query function is not the right place to start from genesis.  It should be initialized upon app init, not in Query.